Add warn-once deprecation warning to legacy sparse constructors

pearu opened the pull request:
Addresses https://github.com/pytorch/pytorch/issues/68323#issuecomment-1425174341

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94850



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer

pearu committed:
Add warn-once deprecation warning to legacy sparse constructors

[ghstack-poisoned]

cpuhrsch commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -553,21 +553,33 @@ Tensor legacy_sparse_tensor_generic_ctor_new(
     at::OptionalDeviceGuard device_guard(deviceOptional);
     return at::sparse_coo_tensor(r.tensor(0), r.tensor(1));
   } else if (r.idx == 3) {
+    TORCH_WARN_ONCE(
```
You could try turning these into errors to see if they trigger CI failures? That could be one way to find hidden invocations deep within the stack that'd result in annoying warnings to users.

pearu replied:
@cpuhrsch 
```
grep -r -I --exclude "*~" 'sparse[.]\w*Tensor' test/
```
reveals 52 usages of `torch.sparse.SparseTensor` while in `torch/` there are none.
So, turning warn-once to an error will guarantee CI failures.

There exist legacy tests that test `SparseTensor` constructors. Do we want to remove these tests?

cpuhrsch replied:
@pearu - yes if we're getting rid of these constructors let's also update those tests. Maybe we'll find a gap in functionality or uncover a bug.

pearu replied:
After replacing warn-once with an error, CI still became green. So, no uncovered bugs here.

cpuhrsch reviewed (changes requested):
Requesting changes so it doesn't get accidentally merged. We don't want these warnings to be triggered deep in the stack by other operations.

pearu committed:
Update on "Add warn-once deprecation warning to legacy sparse constructors"

Addresses https://github.com/pytorch/pytorch/issues/68323#issuecomment-1425174341




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse.py:
```diff
@@ -2531,11 +2526,11 @@ def test_sparse_add_coalesce(self, device, dtype):
-    def test_storage_not_null(self):
-        x = torch.cuda.sparse.FloatTensor(2)
+    def test_storage_not_null(self, device):
+        x = torch.empty(2, dtype=torch.float32, layout=torch.sparse_coo, device=device)
```
Let's use the sparse constructors. I think this is supposed to go away @amjames?

amjames replied:
Yes, I am working on a change that will deprecate user facing `empty` and `empty_like` for sparse layouts. 

You should use the constructors directly, or `zeros` would be the closest replacement here.

pearu replied:
On second thought, we can also use `sparse_coo_tensor` constructor:
```python
x = torch.sparse_coo_tensor((2,), dtype=torch.float32)
```
which is the shortest variant among other options.

pearu committed:
Update on "Add warn-once deprecation warning to legacy sparse constructors"

Addresses https://github.com/pytorch/pytorch/issues/68323#issuecomment-1425174341




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Add warn-once deprecation warning to legacy sparse constructors"

Addresses https://github.com/pytorch/pytorch/issues/68323#issuecomment-1425174341




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu committed:
Update on "Add warn-once deprecation warning to legacy sparse constructors"

Addresses https://github.com/pytorch/pytorch/issues/68323#issuecomment-1425174341




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse.py:
```diff
@@ -82,6 +84,59 @@ def ignore_op(func):
         )
 
+class TestSparse000(TestCase):
+    # we want to run the tests below once before all other test cases,
```
I don't think we can guarantee this order on some internal systems. I assume that's only important for development of this PR? If so can we change the name to something more expressive?

pearu replied:
This test only works as intended when it is run before other tests that call legacy constructors because of the warn-once nature. That said, I now realize that other tests should never call the legacy constructors anyway, and if these do, these should be fixed. In fact, the test would be even stronger when the test is run after all other tests, although, I think we can consider it unnecessary.

In sum, I'll just rename the test class to TestSparseLegacyConstructors.

pearu replied:
Fixed.

cpuhrsch commented on test/test_sparse.py:
```diff
@@ -82,6 +84,59 @@ def ignore_op(func):
+            with redirect_stderr(cerr):
+                test_f()
+                test_f()  # checks warns once
+            FileCheck().check_count(str=test_f.__doc__, count=1, exactly=True).run(cerr.getvalue())
```
You might be able to use [`assertWarns`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarns).

pearu replied:
yes, thanks for the hint!

Btw, we have `assertWarnsOnceRegex` in torch/testing/_internal/common_utils.py that naming is misleading: it calls `set_warn_always(True)` to force `TORCH_WARN_ONCE` to become `TORCH_WARN` but that behavior is contradictory to the method name --- I would expect that it would implement a true test for the "warn once" feature.

cpuhrsch commented on test/test_sparse.py:
```diff
@@ -947,7 +996,7 @@ def test_not_in_place(x):
         def test_shape(sparse_dims, nnz, sizes):
             x, _, _ = self._gen_sparse(sparse_dims, nnz, sizes, dtype, device, coalesced)
-            zeros = torch.zeros(sizes, layout=torch.sparse_coo).to(x.device)
+            zeros = torch.sparse_coo_tensor(sizes).to(x.device)
```
nit: could pass device directly to constructor

pearu replied:
right, fixed.

cpuhrsch commented on test/test_sparse.py:
```diff
@@ -2821,54 +2849,33 @@ def test_tensor(indices, values, indices_equal, values_equal):
         s = torch.sparse_coo_tensor(i, v, size)
 
         self.assertEqual(torch.sparse_coo, s.new(device='cpu').layout)
-        self.assertRaises(TypeError, lambda: s.new(v.storage()))
```
Hm, why this change?

pearu replied:
It resolves the following warning:
```
UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
```

cpuhrsch commented on test/test_sparse.py:
```diff
@@ -2880,13 +2887,46 @@ def test_dtypes(self, device):
+
+        self._test_empty_full(device, dtype, requires_grad)
         if torch.cuda.device_count() > 0:
-            do_test_empty_full(self, all_sparse_dtypes, torch.sparse_coo, None)
```
Does this mean we can delete `do_test_empty_full`?

pearu replied:
Not yet. `TestTensorCreation.test_empty_full` in test/test_tensor_creation_ops.py needs to be updated as well. It is slightly OT for this PR.

cpuhrsch replied:
It seems safer to update `do_test_empty_full` with support for sparse layouts than creating another copy of it. Why not extend `do_test_empty_full`?

pearu replied:
@cpuhrsch according to
https://github.com/pytorch/pytorch/blob/5f24b2b1f062e930c41b576e44195f9912dbc958/torch/testing/_internal/common_utils.py#L3676-L3677
the `do_test_empty_full` function will be dropped. So, this PR applies the FIXME to sparse layouts.

cpuhrsch reviewed (changes requested):
Looking great! Just seen my inline comments.

pearu committed:
Update on "Add warn-once deprecation warning to legacy sparse constructors"

Addresses https://github.com/pytorch/pytorch/issues/68323#issuecomment-1425174341




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

cpuhrsch reviewed (approved):
Looks good, see last comment about `do_test_empty_full`.

Thank you!

cpuhrsch commented:
@pytorchbot merge

pearu commented:
@pytorchbot merge

pearu committed:
Update on "Add warn-once deprecation warning to legacy sparse constructors"

Addresses https://github.com/pytorch/pytorch/issues/68323#issuecomment-1425174341




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu commented:
@pytorchbot merge
